File: 992-subarrays-with-k-different-integers/992-subarrays-with-k-different-integers.java
class Solution {
    public int subarraysWithKDistinct(int[] nums, int k) {
        return atMostK(nums, k) - atMostK(nums, k - 1);
    }
    public int atMostK(int[] nums, int k) {
        int N = nums.length;
        int res = 0, left = 0, cnt = 0;
        Map<Integer, Integer> map = new HashMap<>();
        for(int i = 0; i < N; i++) {
            //put in
            map.put(nums[i], map.getOrDefault(nums[i], 0) + 1);
            //move out
            while(map.size() > k) {
                map.put(nums[left], map.getOrDefault(nums[left], 0) -1);
                if(map.get(nums[left]) == 0)
                    map.remove(nums[left]);
                left++;
            }
            //calculate
            //***a little bit ambiguous/hard to understand***
            res += i - left + 1;
        }
        return res;
    }
}

File: 992-subarrays-with-k-different-integers/992-subarrays-with-k-different-integers.py
class Solution:
    def subarraysWithKDistinct(self, nums: List[int], k: int) -> int:
        
        def atMost(nums: List[int], k : int):
            l, n, res= 0, len(nums), 0
            dic = defaultdict(int)
            for i in range(n):
                if dic[nums[i]] == 0: k -= 1
                dic[nums[i]] += 1
                while k < 0:
                    dic[nums[l]] -= 1
                    if dic[nums[l]] == 0: k += 1                    
                    l += 1
                res += i - l + 1
            return res
        
        return atMost(nums, k) - atMost(nums, k - 1)
        
                